Feature: Add Folia Support

Quagington opened the pull request:
This change introduces a library called [FoliaLib](https://github.com/TechnicallyCoded/FoliaLib) that allows this plugin to run on both Folia and Spigot/Paper. This library includes methods to replace scheduling calls.

Most of the changes in this PR are agnostic inline scheduler call replacements. However, for uses of BukkitRunnable a more intensive refactor was required as there is not an equivalent in Folia/FoliaLib.

LOOHP commented:
I'm not sure if just replacing all the schedulers would work, as there isn't a main thread anymore. Some calls depend on or must be done on the main thread, so those ones might be a bit more difficult to get right on Folia.

Quagington commented:
> I'm not sure if just replacing all the schedulers would work, as there isn't a main thread anymore. Some calls depend on or must be done on the main thread, so those ones might be a bit more difficult to get right on Folia.

<img width="823" alt="image" src="https://github.com/user-attachments/assets/1d166136-68b9-48e3-ace0-60e639adfd08" />

There still is the concept of a "Global" scheduler. For other things that would modify inventories, entities, etc. those would need to be done at a specific location and I tried to account for that in the PR. I recognize that there is heavy testing overhead with this PR and would be happy to assist however I can.

Many servers are moving to Folia and many plugins are adding support.  Would love to help out but if not it's totally up to you. Cheers.

LOOHP commented:
Yes, I would love to support Folia, and support Folia for all my remaining plugins as well. The main thing standing in the way of merging this PR is just the testing part as I don't have that much time to research and test Folia just yet. But thank you for your work and rest assured that it will definitely be done at some point.

Quagington commented:
I understand Folia might not be a priority and this change would require testing - however, many large servers are going to be supporting Folia and Paper has been spending increasing time on it. Hopefully this can at least be a starting point - I am happy to help anyway I can.

LOOHP commented:
Actually, with #256, I plan to merge that one first and then perhaps get folia support to work after those changes. Therefore it might be best to not make the changes first.
I hope to include both features for the 1.21.5 update.

Quagington commented:
Hi, I'm glad this is getting some attention again. I've been using this on my server for months with no issues! I'll see if I can address your feedback soon.

Quagington commented on common/src/main/java/com/loohp/interactivechat/Commands.java:
```diff
@@ -87,7 +87,7 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
                 InteractiveChat.placeholderCooldownManager.reloadPlaceholders();
                 PlayerUtils.resetAllPermissionCache();
-                Bukkit.getScheduler().runTaskAsynchronously(InteractiveChat.plugin, () -> InteractiveChat.playerDataManager.reload());
+                InteractiveChat.plugin.getScheduler().runAsync((task) -> InteractiveChat.playerDataManager.reload());
```
Fixed

Quagington commented on common/src/main/java/com/loohp/interactivechat/hooks/eco/EcoHook.java:
```diff
@@ -67,10 +67,16 @@ public static Future<ItemStack> setEcoLore(ItemStack itemStack) {
+        if (!InteractiveChat.ecoSetLoreOnMainThread || plugin.getScheduler().isGlobalTickThread()) {
             return CompletableFuture.completedFuture(setEcoLore0(itemStack.clone(), player));
         } else {
-            return Bukkit.getScheduler().callSyncMethod(InteractiveChat.plugin, () -> setEcoLore0(itemStack.clone(), player));
```
Agreed this would be nice

Quagington commented on common/src/main/java/com/loohp/interactivechat/objectholders/MentionPair.java:
```diff
@@ -67,7 +67,6 @@ public int hashCode() {
         int result = 1;
         result = prime * result + ((receiver == null) ? 0 : receiver.hashCode());
         result = prime * result + ((sender == null) ? 0 : sender.hashCode());
-        result = prime * result + taskid;
```
It looks like you've since added some sort of timestamp here. Still needed?

Quagington commented on common/src/main/java/com/loohp/interactivechat/Commands.java:
```diff
@@ -359,7 +359,7 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
                     if (sender instanceof Player) {
                         String message = String.join(" ", Arrays.copyOfRange(args, 1, args.length));
-                        Bukkit.getScheduler().runTask(InteractiveChat.plugin, () -> {
+                        InteractiveChat.plugin.getScheduler().runNextTick((task) -> {
```
This one I'm actually not sure of, good question. Any advice on how to test?

Quagington commented:
> Actually, with #256, I plan to merge that one first and then perhaps get folia support to work after those changes. Therefore it might be best to not make the changes I suggested in this PR first. I hope to include both features for the 1.21.5 update.

Looks like you've merged #256! I've pulled your changes in and have addressed some of your comments.

Quagington commented:
Also a note that I have a PR open for the DiscordSRV as well https://github.com/LOOHP/InteractiveChat-DiscordSRV-Addon/pull/69

Quagington commented:
Why was this closed @LOOHP ?
